Evolving contracts - Deep dive (#5435)
Particular/docs.particular.net: nservicebus/messaging/evolving-contracts.md

For complex systems, especially when there are significant changes in a message type, such as adding or removing properties, or changing the property type, a step-by-step approach to evolving message contracts is recommended.

#### Adding data to a contract

Depending on the complexity of the change, the number of senders and receivers for the contract as well as how the change might propagate through the system, a different approach may be desirable. It's important to consider where the added data is required for a message to be successfully processed.

Considering a scenario in which the data that is added to the contract is not required to succesfully process a message:

* Update the contract and release a new version.
  * Instead of relying on .NET to set the default value for int Age = 1, it's better to use nullable types and represent missing values as null. This allows for message handlers to reliably verify if the data is available using the ´HasValue´-property.
* Update senders to use the new contract version.
* Update receivers to handle the new contract version. 

>   * This can be done gradually, endpoint by endpoint. 

Author: ```suggestion
  * This can be done gradually, endpoint by endpoint. Endpoints that are not upgraded and reference the previous contracts assembly, will be unaware of the newly added data as it will be ignored during deserialization.
```

Author: @dvdstelt This was in my notes but I forgot to put it in the PR, do you think it makes sense to add this to further clarify?

Reviewer: You can just merge this, but if you think it needs to be clarified further, then yeah.

Reviewer 2: Sounds good!

For a more complex change in which receivers require the data in order to successfully process a message, a more gradual upgrade is recommended.

For the contracts assembly:

* Create a new contract that either:
  * is a copy of the original contract, adding the additional properties that are needed
  * inherits the previous contract instead of adjusting the same type
* Release a new version of the contract's assembly

For senders and publishers:

* Update senders and publishers to target the new contracts assembly
* Update senders and publishers use the new message contract when sending/publishing messages

> This may be done gradually, deploying endpoint by endpoint as suited.

Author: ```suggestion
This may be done gradually, deploying endpoint by endpoint as suited. Endpoints that are not upgraded and reference the previous contracts assembly, will be unaware of the newly added data as it will be ignored during deserialization.
```

In a next major version, the obsoleted types may be removed.

#### Removing data from a contract

When the need rises to remove data from a contract, the easiest way to implement this change is to start at the receivers' side.

Adjust all receivers to not rely on the data that needs to be removed by:

  * Retrieving the data from somewhere else
  * Removing the data completely

In the first case, the same way of working can be applied as when adding data to a contract:

* Convert the original message handler to initiate a saga
* When the message is received, and the handler identifies that the required data is not part of the incoming message, send a dedicated message to the relevant endpoint to retrieve the missing information. If needed, keep track of the data from the original message by storing all relevant information in the saga
* When the data is retrieved, the original message handler logic may be invoked to process the data

> When all receivershave been updated to allow for processing with less data, the contracts can be adjusted.

Reviewer: ```suggestion
When all receivers have been updated to allow for processing with less data, the contracts can be adjusted.
```

Depending on the complexity of the system, multiple options are possible. The simplest one is to:

* Update the contract to the new version.
* Update senders to use the new contract version. Ensure changes are visible for receivers, for example, by decorating the existing property with the Obsolete attribute when removing or renaming properties.
* Update receivers to handle the new contract version. It's important that the new properties are handled correctly:
  * It's necessary to carefully examine what the default values for the added properties will be, especially if endpoints running in other versions don't recognize them. In particular, it's important to consider how clients might interpret the default value and provide appropriate guidelines for them.
  * Instead of relying on .NET to set the default value for int Age = 1, it's better to use nullable types and represent missing values as null.

For more complex systems, especially when receivers need to be gradually upgraded, a more gradual upgrade is recommended:

* Create a new contract that inherits the previous version instead of adjusting the same type
* Release vX.1 of the contract's assembly

> * Update senders to use the new contract version, this may be done gradually.

Author: ```suggestion
* Update senders to use the new contract version, this may be done gradually, deploying endpoint by endpoint as suited.
```

  * When the actual values for the new properties are required and the data is available in the storage, it can be retrieved as part of the message processing
  * When the new properties are required to correctly process the message and are stored by another service:
    * Convert the original message handler to initiate a saga
    * When the message is received, and the handler identifies that a part of the data is missing, send a dedicated message to the relevant endpoint to retrieve the missing information. If needed, keep track of the data from the original message by storing all relevant information in the saga
    * When the data is retrieved, a new message containing all required information can be sent. This message will be handled by the same handler that processes the new message contract type, deferring the actual processing to a single handler.

> When all senders and receivers are updated and in-flight messages in the old format have been handled, obsolete the previous contract type. By decorating the previous contract type with the `Obsolete` attribute, the changes become visible for receivers when they upgrade to the new version.

Author: ```suggestion
When all senders and receivers are updated and in-flight messages in the old format have been handled, the previous contract type can be obsoleted. By decorating the previous contract type with the `Obsolete` attribute, the changes become visible for receivers when they upgrade to the new version.
```

## Common challenges

### Handling breaking changes in contracts

There are a number of approaches to handling breaking changes in messages. One of the simplest is adding a version to the type name and creating handlers for the new contract. The handlers for the old contract might still be running, to handle in-flight messages and messages generated by old endpoints.

For complex systems, especially when there are significant changes in a message type, such as adding or removing properties, or changing the property type, a step-by-step approach to evolving message contracts is recommended.

#### Adding data to a contract

Depending on the complexity of the system, multiple options are possible. The simplest one is to:

> * Update the contract to the new version.

Author: ```suggestion
* Update the contract and release a new version.
```

* Update senders to use the new contract version. Ensure changes are visible for receivers, for example, by decorating the existing property with the Obsolete attribute when removing or renaming properties.
* Update receivers to handle the new contract version. It's important that the new properties are handled correctly:
  * It's necessary to carefully examine what the default values for the added properties will be, especially if endpoints running in other versions don't recognize them. In particular, it's important to consider how clients might interpret the default value and provide appropriate guidelines for them.

>   * Instead of relying on .NET to set the default value for int Age = 1, it's better to use nullable types and represent missing values as null.

Author: ```suggestion
  * Instead of relying on .NET to set the default value for int Age = 1, it's better to use nullable types and represent missing values as null. That way, a specific check for a missing value can be added, as opposed to checking for a default value that might represent a valid value.
```

>   * Instead of relying on .NET to set the default value for int Age = 1, it's better to use nullable types and represent missing values as null.

Author: @dvdstelt Is this better?

  * When the new properties are required to correctly process the message and are stored by another service:
    * Convert the original message handler to initiate a saga
    * When the message is received, and the handler identifies that a part of the data is missing, send a dedicated message to the relevant endpoint to retrieve the missing information. If needed, keep track of the data from the original message by storing all relevant information in the saga
    * When the data is retrieved, a new message containing all required information can be sent. This message will be handled by the same handler that processes the new message contract type, deferring the actual processing to a single handler.

When all senders and receivers are updated and in-flight messages in the old format have been handled, the previous contract type can be obsoleted. By decorating the previous contract type with the `Obsolete` attribute, the changes become visible for receivers when they upgrade to the new version.
In a next major version, the obsoloted types may be removed.

#### Removing data from a contract

> When the need rises to remove data from a contract, the easiest way to implement this change is to start at the consumers' side.

Author: Maybe it's unclear to me what the terminology is that we use depending on the type of message contract.
Command --> receiver
Event --> subscriber
Message/Command/Event --> Consumer
?

> When the need rises to remove data from a contract, the easiest way to implement this change is to start at the consumers' side.

Author: As discussed I'll sweep through the document to align usage of the terminology

> Adjust all receivers and subscribers to not rely on the data that needs to be removed by:

Author: See my comment ☝️

  * Retrieving the data from somewhere else
  * Removing the data completely

In the first case, the same way of working can be applied as when adding data to a contract:

* Convert the original message handler to initiate a saga
* When the message is received, and the handler identifies that the required data is not part of the contract, send a dedicated message to the relevant endpoint to retrieve the missing information. If needed, keep track of the data from the original message by storing all relevant information in the saga
* When the data is retrieved, the original message handler logic may be invoked to process the data

> When all receivers and subscibers have been updated to allow for processing with less data, the contracts can be adjusted. By decorating the previous contract type with the `Obsolete` attribute, the changes become visible for receivers when they upgrade to the new version.

Author: You can, but only when the handler doesn't rely on the value being there, or you'd run into NullReferenceExceptions very quickly. I think the cases in which you can just delete the property and do absolutely nothing are more exceptions in reality.

The main difference to keep in mind when removing data from a contract, is to start at the receiving end of the messages.

#### Modifying serialization formats

Another approach for handling breaking changes is to modify serialization formats. The step-by-step guidance is provided in the [transition serialization formats](/samples/serializers/transitioning-formats/) formats.

### Breaking down large contract assemblies

In the early days of a system, combining all events, commands and messages into a single contracts assembly might be a good place to start. However, as the system grows, breaking down the contracts into smaller parts makes more sense. An obvious reason is a rise in the number of subscribers for events published by a specific endpoint. It's not desirable to expose commands that are meant to be consumed by a single receiver to all the subscribers interested in that same endpoint's events.

At that point it makes more sense to break down the contracts into multiple assemblies. A possible structure solution is to have multiple assemblies per endpoint with a naming convention similar to:

> * ServiceName.Commands
> * ServiceName.Events
> 
> Or even more strict:
> 
> * ServiceName.AutonomousComponent.Commands
> * ServiceName.AutonomousComponent.Events

Author: ```suggestion
* EndpointName.Commands
* EndpointName.Events

Or even more strict:

* EndpointName.AutonomousComponent.Commands
* EndpointName.AutonomousComponent.Events
```
